### ISSUE TITLE
Zed memory usage greater than vscode and memory leak issue on long run

### ISSUE THREAD

**Issue by sheikalthaf** [zed-industries/zed#7939]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

Zed using more memory than vscode.
**VSCode** - **127MB** for big angular project
**Zed** - **220MB** for basic HTML and ts file project

I left zed and vscode over the night and in the morning i see that zed memory consumption is more (**350MB**).

I think there is a memory leak in Zed and also Why it is consuming more memory than VSCode.

### Environment

Zed: v0.123.1 (Zed Preview)
OS: macOS 14.2.1
Memory: 16 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

![Screenshot 2024-02-17 at 9 45 13 AM](https://github.com/zed-industries/zed/assets/22526247/2f69f0e3-8247-46e0-8339-fde4f60ac255)

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

**Comment by N8th8n8el** [comment]
I also think there is a memory leak somewhere. When I freshly start a pretty large ts project Zed consumes ~280MB, but sometime after a day of work it's over 4GB.

**Comment by SomeoneToIgnore** [comment]
Leak could be caused by the same issue as https://github.com/zed-industries/zed/issues/7438
More contribution to the memory bloat might be caused by the way we store LSP logs: they are capped but maybe not enough, see https://github.com/zed-industries/zed/pull/3134 for more context

**Comment by synecdokey** [comment]
https://github.com/typescript-language-server/typescript-language-server/issues/472#issuecomment-1193381252 is also something that makes VSCode consume less memory with `tsserver` compared to a standalone `typescript-language-server`

**Comment by AidanRS5772** [comment]
Ive been dealing with the same thing. Im primarily working on a Rust project with a couple files with a few thousand lines of code each. When I open Zed it uses 200-300 MB of memory but after not that long period of time, sometimes as short as an hour or two, it will be anywhere between 4GB too 8GB. I end up killing it and reopening after I get the rainbow wheel of death, and then I go for another couple of hours. I would obviously perfer not to do this but it isnt a deal breaker as I really like the editor. It is something the Zed team should definitly fix and quickly.

**Comment by mocenigo** [comment]
Maybe some circular reference that is not properly handled with a Weak reference?

**Comment by mrnugget** [comment]
I've left some instructions on how to capture memory profiles here: https://github.com/zed-industries/zed/issues/8436#issuecomment-2003220816

If someone wants to take a stab at that, that'd be nice. Otherwise it'd be great to have an example project that we can use to reproduce.

**Comment by Korywon** [comment]
Bumping this. I leave my editors open for several days and got to 19 GB. Not even sure how that's possible considering I have 16 GB of RAM on this laptop.

<img width="436" alt="image" src="https://github.com/user-attachments/assets/b6297e5c-1624-4bf3-b9c0-0b81eb0bc9ea">

**Comment by mrnugget** [comment]
> Bumping this. I leave my editors open for several days and got to 19 GB.

What language servers did you have running? Did they show up in the activity monitor?  What's real memory usage?

See the optional columns in Activity Monitor:

![screenshot-2024-09-10-12 05 03@2x](https://github.com/user-attachments/assets/7f1556f9-b8bf-4a24-9cb9-5c63a4b28d80)


> Not even sure how that's possible considering I have 16 GB of RAM on this laptop.

Virtual memory. macOS uses compression, swapping, etc.

**Comment by Korywon** [comment]
Ah, that makes much more sense. I mainly use ruby-lsp and whatever Typescript/TSX is using. I don't think I see them in the monitor, but will check again if I hit the memory warning again.

**Comment by JustBipin** [comment]
i'm new to rust, and well not really experienced in anything else either. wasn't rust supposed to free from memory leaks? I mean that's how it's marketed as c++'s alternative. the memory safe and modern alternative.

**Comment by mocenigo** [comment]
> i'm new to rust, and well not really experienced in anything else either. wasn't rust supposed to free from memory leaks? I mean that's how it's marketed as c++'s alternative. the memory safe and modern alternative.

No. It's memory safe. It is a different thing. For instance, no use-after-free, or double free. But memory leaks can still happen.

**Comment by JustBipin** [comment]
memory leak is "memory not freed" right? well rust does have elegent system of freeing memory,the borrow checker and the scope. Still makes no sense to me.
